packages/ui-extensions-react/src/use-context.ts

```diff
@@ -0,0 +1,117 @@
+import * as uiExtensionsSDK from '@datadog/ui-extensions-sdk';
+import * as React from 'react';
+
+/**
+ * This represents the different states the {@link uiExtensionsSDK.Context} can be in.
+ */
+type Context =
+    | ContextHandshakeFailure
+    | ContextInitialized
+    | ContextInitializing;
+
+/**
+ * There was some kind of failure with the handshake.
+ * There's no useful {@link uiExtensionsSDK.Context}, but there is an error.
+ */
+type ContextHandshakeFailure = { type: 'handshake failure'; error: unknown };
+
+/**
+ * The handshake is still being preformed.
+ * There's no useful {@link uiExtensionsSDK.Context} yet.
+ */
+type ContextInitializing = { type: 'initializing' };
+
+/**
+ * The handshake has succeeded and there is a {@link uiExtensionsSDK.Context} that can be used.
+ */
+type ContextInitialized = {
+    type: 'initialized';
+    context: uiExtensionsSDK.Context;
+};
```

Comment:
I have an alternative schema proposal: Many react developers are familiar with hook patterns for async operations that look like what you end up using from `useQuery`: 

```
const { data, isLoading, error } = `useQuery(...)`;
```

Rather than using an enumerated type, we could try to mimix that pattern: 

```
interface useContextResult {
  isLoading: boolean;
  context: uiExtensionsSDK.Context | null;
  error: Error | null
}

const { isLoading, error, context } = useContext(client);
```

What do you think? Of course it's the same information in the end, just attempting to follow existing patterns more closely

Reply:
**tl;dr;** I'm okay with whatever here. I'll update this next week unless the longer response compels you otherwise. You don't have to read the rest of this, but can if you'd like.

---

I have a more thoughts about this, but this response ended up being long enough as it is. If you'd like me to go into even more detail lemme know.

Whatever the API of this ends up being, there's maybe a more important point to make. I don't see this particular hook being that useful for App developers in their day-to-day work, but very useful for IDX. There's at least two reasons why I say this:
1. The SDK `Context` mostly has optional data, and trying to drill down on that seems harder than it needs to be.
2. I think most App developers would rather use something else aside from `Context`.

The second point I think is where we're going to get the most benefit out of having a package like this. I was playing around with a `useCustomWidgetOptions` hook that built on top of this `useContext` hook. It seemed to remove the majority of the explicit references to the `Context`. I feel like it'd be similar for most other hooks we'd build: `useTimeframe`, `useColorTheme`,  etc. All of these have decent defaults and can plaster over the different states of `useContext` no matter what its API looks like.

`Context` gives a lot of data, and it can be useful to solve a lot of problems we run into. For sure we should make it available to App developers can unblock themselves if we don't provide an easy thing to use. But, it feels more like a low-level primitive than something we expect people to work with in day-to-day. And we should try to capture those patterns that people do instead. I didn't want to add that to the documentation because there's nothing else to point people to at the moment.

To that end, I feel like we should optimize correctness over familiarity. In this situation, familiarity makes it harder to know you're doing the right thing when using `useContext`. There's eight different states to deal with:

| `isLoading` | `context` | `error` |
| ----------- | --------- | ------- |
| `false`     | `null`    | `null`  |
| `false`     | `null`    | `Error` |
| `false`     | `Context` | `null`  |
| `false`     | `Context` | `Error` |
| `true`      | `null`    | `null`  |
| `true`      | `null`    | `Error` |
| `true`      | `Context` | `null`  |
| `true`      | `Context` | `Error` |

Whereas with the discriminated union there's only three actual states:

| `isLoading` | `context` | `error` | Actual states         |
| ----------- | --------- | ------- | --------------------- |
| `false`     | `null`    | `null`  | `"initializing"`        |
| `false`     | `null`    | `Error` |                       |
| `false`     | `Context` | `null`  |                       |
| `false`     | `Context` | `Error` |                       |
| `true`      | `null`    | `null`  |                       |
| `true`      | `null`    | `Error` | `"handshake failure"` |
| `true`      | `Context` | `null`  | `"initialized"`       |
| `true`      | `Context` | `Error` |                       |

For something that seems like it'll end up being more internal than external, it seems like we should optimize for simplifying on our end.

Of course, if my read on `useContext` and the SDK `Context` in general is off, then yeah we should optimize for App developer familiarity.

Reply:
A summary of in-person discussions:

1. I like the pattern of this hook returning `Context | null` to start with
2. If we need extra hooks for handling handshake errors on the developer's end, we can do so later and explicitly
3. I like that we have a data model with shape, it makes stuff discoverable. The data model will come to be more and more complex over time. Ideally there is a way to both have a general and explorable data model and specificity for common usecases in a way that feels layered rather than disjoint. One idea discussed was a context selector pattern: `useContextSelector(getDashboardTimeframe)`. 
4. I like the idea of automatically calling `init` inside this hook or another wrapper. The one usecase it won't cover is defining an auth provider, but this is generally done outside react at present. Later we may move to a setProvider() model anyways

Reply:
I tried to do the fourth thing on the list, calling `init` within the hook. But I can't get the tests to pass in that case. Leads me to one of two ideas: either we need to mock things differently (most likely), or this might be a bug in `jest` that's fixed in a newer version (less likely).

In either case, I'm going to punt on it until we can add some tests for that behavior that pass.